File: src/jimin/week4/광고 삽입.py
def solution(play_time, adv_time, logs):
    play_time = get_time(play_time)
    adv_time = get_time(adv_time)
    total_time = [0] * (play_time + 1)

    for log in logs:
        start, end = log.split("-")
        start_second = get_time(start)
        end_second = get_time(end)

        total_time[start_second] += 1
        total_time[end_second] -= 1

    # 1, -1을 처리하기 위한 구간 합
    for i in range(1, play_time):
        total_time[i] += total_time[i - 1]
    #print(total_time)

    # 사람에 대한 구간 합
    for i in range(1, play_time):
        total_time[i] += total_time[i - 1]

    max_people = -1
    answer = 0  # 시작 위치
    # 끝나는 시간 기준으로 반복
    for i in range(adv_time - 1, play_time):
        if i >= adv_time:
            temp = total_time[i] - total_time[i - adv_time]
        else:
            temp = total_time[i]
        if temp > max_people:
            max_people = temp
            answer = i - adv_time + 1

    return get_str(answer)


def get_time(time):
    h, m, s = time.split(':')
    return int(h) * 60 * 60 + int(m) * 60 + int(s)


def get_str(time):
    h = time // 3600
    time = time % 3600
    m = time // 60
    s = time % 60
    return str(h).zfill(2) + ":" + str(m).zfill(2) + ":" + str(s).zfill(2)


play_time = "02:03:55"
adv_time = "00:14:15"
logs = ["01:20:15-01:45:14", "00:40:31-01:00:00", "00:25:50-00:48:29", "01:30:59-01:53:29", "01:37:44-02:02:30"]

play_time = "99:59:59"
adv_time = "25:00:00"
logs = ["69:59:59-89:59:59", "01:00:00-21:00:00", "79:59:59-99:59:59", "11:00:00-31:00:00"]

play_time = "00:00:10"
adv_time = "00:00:05"
logs = ["00:00:08-00:00:10"]

print(solution(play_time, adv_time, logs))

File: src/taeeun/week4/광고삽입.cpp
#include <string>
#include <vector>
#include <iostream>
#include <algorithm>

using namespace std;
int getHour(string a) {return stoi(a.substr(0, 2));}
int getMin(string a) {return stoi(a.substr(3, 2));}
int getSec(string a) {return stoi(a.substr(6, 2));}

int convertTime(string time) {
    return getHour(time)*3600+getMin(time)*60+getSec(time);
}

string toString(int time) {
    string s;
    
    int hour=time/3600;
    int min=(time/60) %60;
    int sec=time%60;
    
    if (hour < 10) s+='0';
    s+=to_string(hour)+':';
    if (min < 10) s+='0';
    s+=to_string(min)+':';
    if (sec<10) s+='0';
    s+=to_string(sec);
    
    return s;
}

string solution(string play_time, string adv_time, vector<string> logs) {
    string answer = "";
    int play=convertTime(play_time);
    int adv=convertTime(adv_time);
    
    vector<long long> arr(play, 0);
    
    for (int i=0; i<logs.size(); i++) {
        string s=logs[i].substr(0, 8);
        string e=logs[i].substr(9, 8);
        
        int start= convertTime(s);
        int end=convertTime(e);
        
        for (int i=start; i<end; i++) {
            arr[i]++;
        }
        
    }
    
    //슬라이딩 윈도우 사용
    //윈도우의 크기는 adv
    
    //윈도우 초기화
    long long max_sum=0;
    for (int i=0; i<adv; i++) {
        max_sum+=arr[i];
    }
    long long sum=max_sum;
    long long  temp=0;
    for (int i=adv; i<play; i++) {
        sum-=arr[i-adv];
        sum+=arr[i];
        if (sum>max_sum) {
            max_sum=sum;
            temp=i-adv+1;
        }
    }
    return toString(temp);
}
